### PR TITLE
[Fix] Set default value of `ann_file`, `data_root` to `None`

### LINKED ISSUE #794: [Bug] CustomDataset overriding of ann_file not possible if base configuration has an ann_file configured
### Prerequisite

- [X] I have searched [Issues](https://github.com/open-mmlab/mmengine/issues) and [Discussions](https://github.com/open-mmlab/mmengine/discussions) but cannot get the expected help.
- [X] The bug has not been fixed in the latest version(https://github.com/open-mmlab/mmengine).

### Environment

OrderedDict([('sys.platform', 'win32'), ('Python', '3.8.15 (default, Nov 24 2022, 14:38:14) [MSC v.1916 64 bit (AMD64)]'), ('CUDA available', False), ('numpy_random_seed', 2147483648), ('MSVC', 'Microsoft (R) C/C++-Optimierungscompi
ler Version 19.16.27045 für x64'), ('GCC', 'n/a'), ('PyTorch', '1.13.0'), ('PyTorch compiling details', 'PyTorch built with:\n  - C++ Version: 199711\n  - MSVC 192829337\n  - Intel(R) Math Kernel Library Version 2020.0.2 Product Bui
ld 20200624 for Intel(R) 64 architecture applications\n  - Intel(R) MKL-DNN v2.6.0 (Git Hash 52b5f107dd9cf10910aaa19cb47f3abf9b349815)\n  - OpenMP 2019\n  - LAPACK is enabled (usually provided by MKL)\n  - CPU capability usage: AVX2
\n  - Build settings: BLAS_INFO=mkl, BUILD_TYPE=Release, CXX_COMPILER=C:/cb/pytorch_1000000000000/work/tmp_bin/sccache-cl.exe, CXX_FLAGS=/DWIN32 /D_WINDOWS /GR /EHsc /w /bigobj -DUSE_PTHREADPOOL -openmp:experimental -IC:/cb/pytorch_
1000000000000/work/mkl/include -DNDEBUG -DUSE_KINETO -DLIBKINETO_NOCUPTI -DUSE_FBGEMM -DUSE_XNNPACK -DSYMBOLICATE_MOBILE_DEBUG_HANDLE -DEDGE_PROFILER_USE_KINETO, LAPACK_INFO=mkl, PERF_WITH_AVX=1, PERF_WITH_AVX2=1, PERF_WITH_AVX512=1
, TORCH_VERSION=1.13.0, USE_CUDA=0, USE_CUDNN=OFF, USE_EXCEPTION_PTR=1, USE_GFLAGS=OFF, USE_GLOG=OFF, USE_MKL=ON, USE_MKLDNN=ON, USE_MPI=OFF, USE_NCCL=OFF, USE_NNPACK=OFF, USE_OPENMP=ON, USE_ROCM=OFF, \n'), ('TorchVision', '0.14.0')
, ('OpenCV', '4.6.0'), ('MMEngine', '0.3.2')])

### Reproduces the problem - code sample

```python
_base_ = '../../../configs/resnet/resnet50_8xb32_in1k.py'

# model overrides
model = dict(
    backbone=dict(
        init_cfg=dict(
            type='Pretrained',
            checkpoint='https://download.openmmlab.com/mmclassification/v0/resnet/resnet50_8xb32_in1k_20210831-ea4938fc.pth',
            prefix='backbone',
        )
    ),
    head=dict(
        num_classes=5
    )
)

# schedule overrides
train_cfg = dict(max_epochs=2, val_interval=10)

# dataset overrides
dataset_type = 'CustomDataset'
data_preprocessor = dict(
    num_classes=5
)

data_folder = 'custom/models/my_data/data'

train_dataloader = dict(
    dataset=dict(
        type=dataset_type,
        data_root=data_folder,
        data_prefix='train',
        ann_file=None,
    )
)

val_dataloader = dict(
    dataset=dict(
        type=dataset_type,
        data_root=data_folder,
        data_prefix='val',
        ann_file=None
    )
)

test_dataloader = dict(
    dataset=dict(
        type=dataset_type,
        data_root=data_folder,
        data_prefix='test',
        ann_file=None
    )
)
```

### Reproduces the problem - command or script

```shell
python tools/train.py custom/models/my_data/my_data.py --work-dir custom/models/my_data/work_dir
```

### Reproduces the problem - error message

```
Traceback (most recent call last):
  File "C:\Users\xx\miniconda3\envs\xx\lib\site-packages\mmengine\registry\build_functions.py", line 121, in build_from_cfg
    obj = obj_cls(**args)  # type: ignore
  File "c:\users\xx\documents\xx\mmcls\datasets\custom.py", line 171, in __init__
    super().__init__(
  File "c:\users\xx\documents\xx\mmcls\datasets\base_dataset.py", line 92, in __init__
    super().__init__(
  File "C:\Users\xx\miniconda3\envs\car_outside_classification\lib\site-packages\mmengine\dataset\base_dataset.py", line 241, in __init__
    self._join_prefix()
  File "C:\Users\xx\miniconda3\envs\xx\lib\site-packages\mmengine\dataset\base_dataset.py", line 534, in _join_prefix
    if not is_abs(self.ann_file) and self.ann_file:
  File "C:\Users\xx\miniconda3\envs\xx\lib\site-packages\mmengine\utils\path.py", line 113, in is_abs
    if osp.isabs(path) or path.startswith(('http://', 'https://', 's3://')):
  File "C:\Users\xx\miniconda3\envs\xx\lib\ntpath.py", line 63, in isabs
    s = os.fspath(s)
TypeError: expected str, bytes or os.PathLike object, not NoneType
```

### Additional information

1. Overriding the setting `ann_file` in a configuration should be possible as stated in the documentation using `None`
2. imagenet_bs32 as base config, inherited dataset should use CustomDataset Subfolder format
3. The error seems to be in base_dataset line 534 https://github.com/open-mmlab/mmengine/blob/bced7d6ef04c28b029b47587ed4c9920dff544a2/mmengine/dataset/base_dataset.py#L534

It should be

```python
if self.ann_file and not is_abs(self.ann_file):
```

to first check if the value is `None `and then execute `is_abs`.

### PR DESCRIPTION
**Pull request by HAOCHENYE**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Resolve #794.

Set default value of `ann_file`, `data_root` to `None`.

## Modification

Please briefly describe what modification is made in this PR.

## BC-breaking (Optional)

Does the modification introduce changes that break the backward-compatibility of the downstream repos?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
3. If the modification has potential influence on downstream projects, this PR should be tested with downstream projects, like MMDet or MMCls.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

File: mmengine/dataset/base_dataset.py [hunk 1]
```diff
@@ -211,21 +211,32 @@ class BaseDataset(Dataset):
     _fully_initialized: bool = False
 
     def __init__(self,
-                 ann_file: str = '',
+                 ann_file: str = None,
```
Comment (zhouzaida) [inline mmengine/dataset/base_dataset.py:214]:
```suggestion
                 ann_file: Optional[str] = None,
```

File: mmengine/dataset/base_dataset.py [hunk 2]
```diff
@@ -211,21 +211,32 @@ class BaseDataset(Dataset):
...
                  metainfo: Optional[dict] = None,
-                 data_root: str = '',
-                 data_prefix: dict = dict(img_path=''),
+                 data_root: str = None,
```
Comment (zhouzaida) [inline mmengine/dataset/base_dataset.py:216]:
```suggestion
                 data_root: Optional[str] = None,
```

**Review by C1rN09 — DISMISSED** [review]

File: mmengine/dataset/base_dataset.py [hunk 3]
```diff
@@ -222,10 +221,11 @@ def __init__(self,
...
                  lazy_init: bool = False,
                  max_refetch: int = 1000):
-
-        self.data_root = data_root
-        self.data_prefix = copy.copy(data_prefix)
         self.ann_file = ann_file
+        self._metainfo = self._load_metainfo(copy.deepcopy(metainfo))
+        self.data_root = data_root
+        data_prefix = copy.copy(data_prefix)
+        self.data_prefix = data_prefix
```
Comment (zhouzaida) [inline mmengine/dataset/base_dataset.py:228]: Why do we need to split `self.data_prefix = copy.copy(data_prefix)` to two lines?

File: mmengine/dataset/base_dataset.py [hunk 4]
```diff
@@ -222,10 +221,11 @@ def __init__(self,
                  test_mode: bool = False,
                  lazy_init: bool = False,
                  max_refetch: int = 1000):
-
-        self.data_root = data_root
-        self.data_prefix = copy.copy(data_prefix)
         self.ann_file = ann_file
+        self._metainfo = self._load_metainfo(copy.deepcopy(metainfo))
```
Comment (zhouzaida) [inline mmengine/dataset/base_dataset.py:225]: Why do we need to move this line to here?

File: mmengine/dataset/base_dataset.py [hunk 5]
```diff
@@ -531,15 +530,15 @@ def _join_prefix(self):
         """
         # Automatically join annotation file path with `self.root` if
         # `self.ann_file` is not an absolute path.
-        if not is_abs(self.ann_file) and self.ann_file:
-            self.ann_file = osp.join(self.data_root, self.ann_file)
+        if self.ann_file and not is_abs(self.ann_file):
+            self.ann_file = osp.join(self.data_root or '', self.ann_file)
```
Comment (zhouzaida) [inline mmengine/dataset/base_dataset.py:534]:
```suggestion
        if self.ann_file and not is_abs(self.ann_file) and self.data_root:
            self.ann_file = osp.join(self.data_root, self.ann_file)
```

File: mmengine/dataset/base_dataset.py [hunk 6]
```diff
@@ -531,15 +530,15 @@ def _join_prefix(self):
...
         # Automatically join data directory with `self.root` if path value in
         # `self.data_prefix` is not an absolute path.
         for data_key, prefix in self.data_prefix.items():
             if isinstance(prefix, str):
                 if not is_abs(prefix):
                     self.data_prefix[data_key] = osp.join(
-                        self.data_root, prefix)
+                        self.data_root or '', prefix)
```
Comment (zhouzaida) [inline mmengine/dataset/base_dataset.py:541]:
```python
                if not is_abs(prefix) and self.data_root:
                    self.data_prefix[data_key] = osp.join(
                        self.data_root, prefix)
```

Comment (HAOCHENYE) [reply mmengine/dataset/base_dataset.py:541]: Even if `data_root=''`, this code should also be executed

File: mmengine/dataset/base_dataset.py (hunk 4 above)
Comment (HAOCHENYE) [reply mmengine/dataset/base_dataset.py:225]: Keep the order of attributes assignment consistent with the order of arguments

**Review by zhouzaida — APPROVED** [review]

**Pull request merged**